File: backend/app/realtime.py
from fastapi import WebSocket
from collections import defaultdict


class ConnectionManager:
    def __init__(self) -> None:
        self.active = defaultdict(list)

    async def connect(self, user_id: str, websocket: WebSocket) -> None:
        await websocket.accept()
        self.active[user_id].append(websocket)

    def disconnect(self, user_id: str, websocket: WebSocket) -> None:
        if user_id in self.active:
            self.active[user_id] = [ws for ws in self.active[user_id] if ws != websocket]

    async def send(self, user_id: str, message: dict) -> None:
        for ws in list(self.active.get(user_id, [])):
            await ws.send_json(message)


manager = ConnectionManager()

File: backend2/src/realtime.ts
import type WebSocket from "ws";

class ConnectionManager {
  active = new Map<string, Set<WebSocket>>();

  connect(userId: string, socket: WebSocket): void {
    const sockets = this.active.get(userId) ?? new Set<WebSocket>();
    sockets.add(socket);
    this.active.set(userId, sockets);
  }

  disconnect(userId: string, socket: WebSocket): void {
    const sockets = this.active.get(userId);
    if (!sockets) return;
    sockets.delete(socket);
    if (!sockets.size) this.active.delete(userId);
  }

  send(userId: string, payload: Record<string, unknown>): void {
    for (const socket of this.active.get(userId) ?? []) {
      if (socket.readyState === socket.OPEN) {
        socket.send(JSON.stringify(payload));
      }
    }
  }
}

export const manager = new ConnectionManager();
